src/components/PivotTable/PivotTableGrid/util.ts

```diff
@@ -0,0 +1,579 @@
+        mixedTableStartRowCache.set(rows[rows.length - 1].path, rowStart)
+      }
+    }
+    const lastKey = mixedTable && cell.key === keys[keys.length - 1]
+    const rowEnd = lastKey ? rowStart + 2 : rowStart + 1
+    const columnEnd = columnStart + columnSpan.value
+    maxRowEnd = max([rowEnd, maxRowEnd])
+    maxColumnEnd = columnEnd > maxColumnEnd ? columnStart : maxColumnEnd
+    const gridArea = new GridArea(rowStart, columnStart, rowEnd, columnEnd)
+
+    if (cell.key === RESULT_PATH_KEY) {
+      const formattedValue = typeof value === 'number' ? formatDecimalOrInteger(value) : value
+      divs.push({
+        types: new Set([PivotTableCellType.VALUE]),
+        gridArea: gridArea,
+        children: formattedValue,
+        isEndRow: mixedTable === undefined,
+      })
+    } else {
+      divs.push({
+        types: new Set([PivotTableCellType.HEADER]),
+        gridArea: gridArea,
+        children: value,
+      })
+    }
+    cellPositions.add(gridArea.toString())
+  }
+  return { maxRowEnd, maxColumnEnd }
+}
+
+function buildVerticalTableHeader<T extends object>(
+  keys: (keyof T)[],
+  rowHeaderSpace: number,
+  divs: PivotTableCellProps[],
+  keysMapping: KeyMap<T>
+) {
+  keys.forEach((k, i) => {
+    const headerGridArea = new GridArea(rowHeaderSpace, i + 1, rowHeaderSpace + 1, i + 2)
+    divs.push({
+      types: new Set([PivotTableCellType.HEADER]),
+      gridArea: headerGridArea,
+      children: keysMapping.get(k).keyName,
+    })
+  })
+}
+
+function buildHorizontalTableHeader<T extends object>(
+  keys: (keyof T)[],
+  columnHeaderSpace: number,
+  divs: PivotTableCellProps[],
+  keysMapping: KeyMap<T>
+) {
+  keys.forEach((k, i) => {
+    const headerGridArea = new GridArea(i + 1, columnHeaderSpace, i + 2, columnHeaderSpace + 1)
+    divs.push({
+      types: new Set([PivotTableCellType.HEADER]),
+      gridArea: headerGridArea,
+      children: keysMapping.get(k).keyName,
+    })
+  })
+}
+
+function buildMixedTableTotals<T extends object>(
+  maxColumnEnd: number,
+  keys: (keyof T)[],
+  mixedTableRowTotals: Map<string, number>,
+  mixedTableColumnTotals: Map<number, number>,
+  totalRowNumber: number,
+  divs: PivotTableCellProps[],
+  cellPositions: Set<string>,
+  mixedTableStartRowCache: Map<string, number>,
+  columnHeaderSpace: number,
+  grandtotal: number
+) {
+  const totalsGridArea = new GridArea(1, maxColumnEnd + 1, keys.length + 2, maxColumnEnd + 2)
+  const dataValueGridArea = new GridArea(totalRowNumber, maxColumnEnd + 1, totalRowNumber + 1, maxColumnEnd + 2)
+
+  mixedTableColumnTotals.forEach(buildMixedTableColumnTotals(totalRowNumber, divs, cellPositions))
+  mixedTableRowTotals.forEach(buildMixedTableRowTotals(mixedTableStartRowCache, maxColumnEnd, divs, cellPositions))
+
+  const gridArea = new GridArea(keys.length + 1, columnHeaderSpace, keys.length + 2, columnHeaderSpace + 1)
+  divs.push({
```

Comment:
it's better to never change values passed as parameters. try to always work with readonly arrays when defining the parameters. This makes the code cleaner and more consistent.